packages/tools/devtools/devtools-view/src/components/graphs/DynamicComposedChart.tsx

```diff
@@ -0,0 +1,394 @@
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { Theme } from "@fluentui/react-components";
+import React, { useState } from "react";
+import {
+	Area,
+	Bar,
+	CartesianGrid,
+	ComposedChart,
+	Legend,
+	Line,
+	ResponsiveContainer,
+	Tooltip,
+	XAxis,
+	YAxis,
+} from "recharts";
+import { ThemeContext } from "../../ThemeHelper";
+
+/**
+ * Data To be rendered with Op Latency Graph
+ */
+export interface GraphDataSet {
+	graphType: "line" | "area" | "bar";
+	schema: {
+		displayName: string;
+		uuid: string;
+		xAxisDataKey: string;
+		yAxisDataKey: string;
+	};
+	data: { [key: string]: number | string }[];
+}
+
+/**
+ * The final shape of the data points passed to the recharts component
+ */
+interface DataPoint {
+	x: string;
+	[key: string]: number | string;
+}
+
+/**
+ * Merges multiple {@link GraphDataSet}'s into singular objects by their y-axis (timestamp) value.
+ * This method is necessary for showing composed graphs beacause Recharts expects data to be in a merged object format
+ */
+const mergeDataSets = (dataSets: GraphDataSet[]): DataPoint[] => {
+	const xAxisDataPointToYAxisDataPointMap: Record<string, Record<string, number | string>> = {};
+
+	for (const dataSet of dataSets) {
+		const { yAxisDataKey, xAxisDataKey, uuid } = dataSet.schema;
+		for (const dataPoint of dataSet.data) {
+			const xAxisDataPoint = dataPoint[xAxisDataKey];
+			xAxisDataPointToYAxisDataPointMap[xAxisDataPoint] = {
+				...xAxisDataPointToYAxisDataPointMap[xAxisDataPoint],
+				[uuid]: dataPoint[yAxisDataKey],
+			};
+		}
+	}
+
+	return Object.keys(xAxisDataPointToYAxisDataPointMap).map((xAxisKey) => {
+		return {
+			x: xAxisKey,
+			...xAxisDataPointToYAxisDataPointMap[xAxisKey],
+		};
+	});
+};
+
+interface Props {
+	/**
+	 * Renders the data as either an Stacked Area or Stacked Bar chart.
+	 * Note that this overrides individually set graphTypes for each dataset.
+	 */
+	stackedGraphType?: "area" | "bar";
+	/**
+	 * The datasets to be rendered onto the chart
+	 */
+	dataSets: GraphDataSet[];
+	/**
+	 * The unit that will be displayed on the y axis
+	 */
+	yAxisUnitDisplayName?: string;
+}
+
+/**
+ * Creates a palette of colors to be used by the DynamicComposedChart.
+ * Because this is using colors from the Fluent ui theme object,
+ * they will automatically update based on the selected theme mode.
+ * The one exception is that for high contrast we override the graph colors because
+ * Fluent defaults many of them to the same color.
+ *
+ * High contrast colors sourced from Fluent Ui React color palette
+ * https://react.fluentui.dev/?path=/docs/theme-colors--page
+ */
+const createGraphColorPalette = (
+	themeMode: string,
+	theme: Theme,
+): {
+	gray: string;
```

Comment:
I can't quite tell from the code if `gray` is used for the chart background, or for actual gray text. If it's the former, should we call it that instead? And also if that's the case, can we find tokens related to `Background` instead of `Foreground` that work here? So if we later do overrides in other places with the intention of changing text color, we don't risk updating something's background too because we overloaded the tokens.

Reply:
I've updated the object to be more clear about how/where the colors are being used in addition to using the gray variants more correctly

Reply:
Completely clear now, thanks 👍